cli/azd/pkg/project/service_target_aks.go

```diff
@@ -0,0 +1,444 @@
+	containerService azcli.ContainerServiceClient
+	az               azcli.AzCli
+	docker           docker.Docker
+	kubectl          kubectl.KubectlCli
+	clock            clock.Clock
+}
+
+func (t *aksTarget) RequiredExternalTools() []tools.ExternalTool {
+	return []tools.ExternalTool{t.docker, t.kubectl}
+}
+
+func (t *aksTarget) Deploy(
+	ctx context.Context,
+	azdCtx *azdcontext.AzdContext,
+	path string,
+	progress chan<- string,
+) (ServiceDeploymentResult, error) {
+	// Login to AKS cluster
+	namespace := t.getK8sNamespace()
+	clusterName, has := t.env.Values[environment.AksClusterEnvVarName]
+	if !has {
+		return ServiceDeploymentResult{}, fmt.Errorf(
+			"could not determine AKS cluster, ensure %s is set as an output of your infrastructure",
+			environment.AksClusterEnvVarName,
+		)
+	}
+
+	// Login to container registry.
+	loginServer, has := t.env.Values[environment.ContainerRegistryEndpointEnvVarName]
+	if !has {
+		return ServiceDeploymentResult{}, fmt.Errorf(
+			"could not determine container registry endpoint, ensure %s is set as an output of your infrastructure",
+			environment.ContainerRegistryEndpointEnvVarName,
+		)
+	}
+
+	log.Printf("getting AKS credentials %s\n", clusterName)
+	progress <- "Getting AKS credentials"
+	credentials, err := t.containerService.GetAdminCredentials(ctx, t.scope.ResourceGroupName(), clusterName)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed retrieving cluster admin credentials, %w", err)
+	}
+
+	kubeConfigManager, err := kubectl.NewKubeConfigManager(t.kubectl)
+	if err != nil {
+		return ServiceDeploymentResult{}, err
+	}
+
+	kubeConfig, err := kubectl.ParseKubeConfig(ctx, credentials.Kubeconfigs[0].Value)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed parsing kube config: %w", err)
+	}
+
+	if err := kubeConfigManager.SaveKubeConfig(ctx, clusterName, kubeConfig); err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed saving kube config: %w", err)
+	}
+
+	if err := kubeConfigManager.MergeConfigs(ctx, "config", "config", clusterName); err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed merging kube configs: %w", err)
+	}
+
+	if _, err := t.kubectl.ConfigUseContext(ctx, clusterName, nil); err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed using kube context '%s', %w", clusterName, err)
+	}
+
+	kubeFlags := kubectl.KubeCliFlags{
+		Namespace: namespace,
+		DryRun:    "client",
+		Output:    "yaml",
+	}
+
+	progress <- "Creating k8s namespace"
+	namespaceResult, err := t.kubectl.CreateNamespace(
+		ctx,
+		namespace,
+		&kubectl.KubeCliFlags{DryRun: "client", Output: "yaml"},
+	)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed creating kube namespace: %w", err)
+	}
+
+	_, err = t.kubectl.ApplyPipe(ctx, namespaceResult.Stdout, nil)
+	if err != nil {
+		return ServiceDeploymentResult{}, fmt.Errorf("failed applying kube namespace: %w", err)
```

Comment:
A k8s namespace is similar to a resource group in azure.  It is primarily there to group a set of related resources within the cluster.

If a cluster was used by multiple teams/products it would be a very common practice to deploy resources into namespaces that have some sort of logical group to avoid naming collisions and other similar concerns.

Reply:
If a namespace is not specified resources will be deployed to a `default` namespace which can get cluttered very easily and mixed with system level resources.

Reply:
but what does it means `applyPipe` ?  Does it mean, assign the namespace after the resource is created?  Like `assign namespace` ?

Reply:
I named it this because it performs a `kubectl apply -f -` but uses `stdin` for the resource to apply.

Reply:
well, the pipe is not really relevant for naming then.
It only redirects the yaml output to the `apply` command.

it can be done without a pipe as:

```
kubectl create namespace foo --dry-run=client -o yaml
kubectl apply -f yamlFile
```

Reply:
I'd rather not write an intermediate file.  I'm open to changing the name but like that it accepts input from stdin for this case.

Reply:
Does it make sense to log the content of the yaml?  

Or, if we need to always call `apply` after create, does it make sense to upgrade our commandRunner to support `piping` ?  (not sure if it is already supported by passing "|" within the list of arguments).